### PR TITLE
feat: add `get_in_memory_or_storage_by_block` to `BlockchainProvider2`

### PR DESCRIPTION
**Pull request by joshieDo**
* Refactor logic into a single method, similar to others get_in_memory_or_storage methods
* fixes race condition on `BlockchainProvider2::history_by_block_hash`

### CODE REVIEW

File: crates/storage/provider/src/providers/blockchain_provider.rs [hunk 1]
```diff
@@ -294,6 +294,32 @@ impl<N: ProviderNodeTypes> BlockchainProvider2<N> {
 
         Ok(None)
     }
+
+    /// Fetches data from either in-memory state or storage by [`BlockHashOrNumber`].
+    fn get_in_memory_or_storage_by_block<S, M, R>(
+        &self,
```
Comment (Rjected) [inline crates/storage/provider/src/providers/blockchain_provider.rs:300]: Would like if the comment referred to "persistent storage", I first interpreted this as EVM storage for some reason, even though it doesn't make sense to interpret it that way

File: crates/storage/provider/src/providers/blockchain_provider.rs [hunk 2]
```diff
@@ -294,6 +294,32 @@ impl<N: ProviderNodeTypes> BlockchainProvider2<N> {
...
+            BlockHashOrNumber::Number(block_number) => {
+                self.canonical_in_memory_state.state_by_number(block_number)
+            }
+        };
+
+        if let Some(block_state) = block_state {
+            return fetch_from_block_state(block_state)
+        }
+        fetch_from_db(self.database_provider_ro()?)
+    }
```
Comment (Rjected) [inline crates/storage/provider/src/providers/blockchain_provider.rs:322]: This makes me wonder if we should just have a method that does the:
```
        if let Some(block_state) = block_state {
            return fetch_from_block_state(block_state)
        }
        fetch_from_db(self.database_provider_ro()?)
```
part, and provide `by_block_hash` or `by_block_num` methods so we don't have to go from number -> id -> match to number. same for hash -> id -> hash

**Review by Rjected — CHANGES_REQUESTED** [review]
this is nice, I have one comment nit and one question about if we should add to the API.

**Review by mattsse — APPROVED** [review]
lgtm

File: crates/storage/provider/src/providers/blockchain_provider.rs (hunk 2 above)
Comment (Rjected) [reply crates/storage/provider/src/providers/blockchain_provider.rs:338]: makes sense

**Review by Rjected — APPROVED** [review]

**Review by fgimenez — APPROVED** [review]

**Pull request merged**